osu.Game/Beatmaps/ControlPoints/TimingControlPoint.cs

```diff
@@ -105,5 +110,35 @@ public bool Equals(TimingControlPoint? other)
                && BeatLength.Equals(other.BeatLength);
 
         public override int GetHashCode() => HashCode.Combine(base.GetHashCode(), TimeSignature, BeatLength, OmitFirstBarLine);
+
+        public List<HitObject> HitObjectsInTimingRange(IBeatmap beatmap)
+        {
+            // If the first group, we grab all hitobjects prior to the next, if the last group, we grab all remaining hitobjects
+            double startTime = beatmap.ControlPointInfo.TimingPoints.Any(x => x.Time < Time) ? Time : double.MinValue;
+            double endTime = beatmap.ControlPointInfo.TimingPoints.FirstOrDefault(x => x.Time > Time)?.Time ?? double.MaxValue;
+
+            return beatmap.HitObjects.Where(x => Precision.AlmostBigger(x.StartTime, startTime) && Precision.DefinitelyBigger(endTime, x.StartTime)).ToList();
+        }
+
+        public void AdjustHitObjectOffset(IBeatmap beatmap, double adjust)
+        {
+            foreach (HitObject hitObject in HitObjectsInTimingRange(beatmap))
+            {
+                hitObject.StartTime += adjust;
+            }
+        }
+
+        public void SetHitObjectBPM(IBeatmap beatmap, double oldBeatLength)
+        {
+            foreach (HitObject hitObject in HitObjectsInTimingRange(beatmap))
+            {
+                double beat = (hitObject.StartTime - Time) / oldBeatLength;
+
+                hitObject.StartTime = (beat * BeatLength) + Time;
+
+                if (hitObject is not IHasRepeats && hitObject is IHasDuration hitObjectWithDuration)
+                    hitObjectWithDuration.Duration *= BeatLength / oldBeatLength;
+            }
+        }
```

Comment:
- `HitObjectsInTimingRange()` requires an external beatmap to do anything useful - at that point it should be a static helper method that takes a beatmap and a control point (can be _any_ control point btw, not even timing), and is also not called by anything public
- `AdjustHitObjectOffset()` doesn't use `this` once except implicitly via `HitObjectsInTimingRange()`
- `AdjustHitObjectOffset()` uses a `double adjust` while `SetHitObjectBPM()` requires being given `oldBeatLength`

None of this is making very much sense as instance methods on `TimingControlPoint`. These all should go to like a static helper class somewhere in `osu.Game.Screens.Edit.Timing`.

Reply:
All but your third point should be addressed - for the third, there's some ways I see to address this, but due to the adjust method needing to be applied before an actual time change occurs, and the set bpm being required to be after a change occurs, not sure what is ideal.

The way I would handle it would be as follows:
I can have them instead take two times for the start and end of a range to get hitobjects, adding a `GetRangeEnd()` function for a `TimingControlPoint`, so the range is just the start and end time of a timing control point, and reworking `HitObjectsInTimingRange` to function with this.

With this, I'd be able to change the `double adjust` to instead be `double oldTime`.